File: integration_python/AvailityCoverages.py
import requests
import json
from AvailityAbstract import AvailityABC


class Coverages(AvailityABC):
    """Provides access to the Health Transactions Coverages API. 
    Can be used to view plan benefits and current status.
    """

    def __init__(self, key, secret):
        super().__init__(key, secret)

    def get(self, parameters: json, returnIds=False):
        """Poll for coverage information given a patients information

        Args:
            returnIds (bool): if returning a list of the coverage search Ids (True)
            or the coverage details (false). Defaults to False. 

            _patient (json | string | dict), optional: json or filename for patient if not using 
            default set for class. 

        Returns:
            list[str] | json: returns list of ids (strings) or a list of coverages in json format. 

        Raises:
            Exception: KeyError if missing required arg for polling 
            config (payer.id and requestTypeCode)
            Exception: If response from coverage API failed
        """
        parameters = self.parseInfo(parameters)
        parameters={x:parameters[x] for x in parameters.keys() if parameters[x] is not None}

        # TODO: uncomment
        # try:
        #     self.checkRequiredArgs(parameters, parameters['payer']['id'], 207)
        # except KeyError as err:
        #     raise Exception('Missing key search parameter(payer id or request type code)') from err

        # TODO: uncomment
        try:
            coveragePoll = requests.post(
                url='https://api.availity.com/availity/v1/coverages',
                headers={
                    'Authorization': self.authentication,
                    'Content-Type': 'application/x-www-form-urlencoded'
                },
                data=parameters
            ).json()
            coverages = coveragePoll['coverages']
        except:
            raise Exception(f"Error polling coverage information\n"
                  f"{coveragePoll['error']: coveragePoll['error_description']}")

        ids = [''] * len(coverages)
        for i in range(len(coverages)):
            ids[i] += coverages[i]['id']

        if returnIds: 
            return ids
        
        return self.getCoveragesSearch(ids)

    def getCoveragesSearch(self, ids):
        """Gets coverage details based on search Ids (from coveragePolling())

        Args:
            ids (list[str])

        Returns:
            json: format "coverages":[list of coverage details per ids]
        """

        results = {'coverages': []}
        for id in ids:
            searchPoll = requests.get(
                url=f'https://api.availity.com/availity/v1/coverages/{id}',
                headers={'Authorization': self.authentication},
            ).json()
            results['coverages'].append(searchPoll)

        return results

File: integration_js/AvailityCoverages.js
import axios from 'axios';

export const url = "/v1/coverages";

/**
 * Polls the coverage API given the patient information (memberID, DOB, State). 
 * @param {bool} returnIds If returning search result Ids (details can be returned from getCoverageSearch).
 *  Otherwise returns full coverage details.
 * @param {dict} patientInfo Specify patients information. Defaults to patient.json.
 * @returns {list} A list of the plans a patient by their ids or the coverage details.
 */

export async function coveragePolling(parameters, returnIds=false) {
    try {
        const poll = await axios.post(
            url,
            parameters,
            { headers: { "Content-Type": "application/x-www-form-urlencoded" } }
        );
        
        let coverages = poll.data['coverages'];
        const ids = [];

        for (let i = 0; i < coverages.length; i++) {
            ids.push(coverages[i]['id']);
        };

        if (returnIds) {return ids;}
        else {
            return await getCoverageSearch(ids);
        };

    } catch(err) {
        console.error('Error fetching data:', err);
    };
};

/**
 * Coverage Details from ids.
 * @param {list} ids ids returned from coveragePolling. 
 * @returns A list of coverage details.
 */

export async function getCoverageSearch(ids) {
    const results = [];
    for (let i = 0; i < ids.length; i++) {
        try {
            const search = await axios.get(
                url + `/${ids[i]}`,
                {},
            );
            results.push(search.data);
        } catch(err) {
            console.error('Error fetching data:', err);
        }
    };
    return results;
};